### DISCUSSION TITLE
Add Lua Hooks

### DISCUSSION THREAD

**Discussion by mmx33** [neomutt/neomutt#4504] [Lua]
Currently NeoMutt offers limited scripting support with Lua. The Lua API is limited to setting and getting variables, and running Mutt Commands. My idea would be to introduce a Hooks API, it differs a bit from the hooks that we all know from Mutt.

In case of NeoMutt a hook cannot determine dynamically to what exact object it refers to, it's all defined by a regex, etc... In Lua we could extend this so that it takes a real object we can act on, such as the maildir, emai or the currently active element of the index and then determines what to do with it.

This could be the first step in making the way for fully-fledged Lua plugins in the spirit of Neovim.

Here's an example of how hypothetical hooks API might look like:

```lua
function onNewMailReceived(mail)
    -- do something with mail
end

function onIndexElementChanged(el)
    folder = mutt.get("folder")
    accounts = {
        string.format("%s/Xyz", folder)
        string.format("%s/Abc", folder)
    }

    for i, acc in ipairs(accounts) do
        if starswith(el.path, acc) then
            mutt.set("spoolfile", string.format("%s/Sent", acc))
        end
    end
end

-- NeoMutt registers a new hook for new-email-received and returns a hook handle
hk = mutt.hook("new-mail-received", onNewMailReceived)
mutt.hook("index-element-changed", onIndexElementChanged)

-- The hook handle can be used to unregister a hook from NeoMutt
mutt.unhook(hk)

-- The hook handle can also be used to modify the currently active hook options
thk = mutt.hook("timer", doSomethingEveryFiveSeconds, { time=5 })
mutt.modhook(thk, { time=10 })
```

**Comment by flatcap** [comment]
NeoMutt's code is a lot better organised than when we first implemented Lua, so this is probably more practical now.

The original issue contains lots more ideas: #153
